pageserver/src/page_cache.rs

```diff
@@ -643,15 +621,10 @@ impl PageCache {
 
     /// Remember that WAL has been received and added to the page cache up to the given LSN
     pub fn advance_last_valid_lsn(&self, lsn: Lsn) {
-        let mut shared = self.shared.lock().unwrap();
+        let oldlsn = self.last_valid_lsn.advance_max(lsn);
 
         // Can't move backwards.
-        let oldlsn = shared.last_valid_lsn;
-        if lsn >= oldlsn {
-            shared.last_valid_lsn = lsn;
-            self.last_valid_lsn.store(lsn);
-            self.seqwait_lsn.advance(lsn);
-        } else {
+        if lsn < oldlsn {
             warn!(
                 "attempted to move last valid LSN backwards (was {}, new {})",
                 oldlsn, lsn
```

Comment:
I don't think I understand the reason for `advance_max`. Is it just because you want the `warn!` instead of the `panic!` inside `SeqWait`?

Reply:
I wonder whether this might be called `try_advance` instead. That name makes me think of returning `Result<(), Lsn>`, which I find amusing but is probably a bit obfuscated.

Reply:
> I don't think I understand the reason for `advance_max`. Is it just because you want the `warn!` instead of the `panic!` inside `SeqWait`?

Yeah, we had a `warn!` there before, I didn't want to change it as part of this patch. But maybe we should just change `advance` to return a Result instead of panicing.

> I wonder whether this might be called try_advance instead. That name makes me think of returning Result<(), Lsn>, which I find amusing but is probably a bit obfuscated.

I chose `advance_max` to follow the nomenclature of Atomic types, where it's called `fetch_max`. Maybe I should've called it `fetch_max` here too, but I like to have the word 'advance' there.

I think I'll remove `advance_max' and make `advance` return a Result instead.